### ISSUE TITLE
Bot placing buy orders when "Buy: Trading Enabled" is unchecked

### ISSUE THREAD

**Issue by chopeta** [chrisleekr/binance-trading-bot#54]
First of all, thanks for sharing this. Great work. I am loving it so far...
I've using the bot for 2 o 3 days now...
Took me some time to understand how the set triggers and %s properly for buy and sell, but I think I've finally figured out.

## Description

I've disabled "Buy: Trading Enabled" on settings which is displaying correcting on the GUI
However, during yesterday BTC dip the bot started buying all over when it wasn't supposed to do it.
So, the bot didn't respect the configuration.
I had to set "Buy: Trigger percentage" to 0.99 to avoid getting buy orders created, which partially fix the problem but not without problems, because when buy is triggered the bot get hanged on that pair and never moves forward until you reboot the docker. 

## To Reproduce

Not sure how to reproduce, but I know is still happening on my install.

## Expected Behaviours

The bot shouldn't create any buy order when "Buy: Trading Enabled" is unchecked.

## Screenshots

![image](https://user-images.githubusercontent.com/7503721/112541040-e7401900-8d80-11eb-807d-c189165cd0a2.png)
![image](https://user-images.githubusercontent.com/7503721/112541056-ee672700-8d80-11eb-87f6-f322c44de2b1.png)
![image](https://user-images.githubusercontent.com/7503721/112541094-f921bc00-8d80-11eb-805e-b2e83cee62cf.png)

## Additional context

My setup is running inside a Ubuntu Server 20.04 KVM machine on a Proxmox server with stable and fast internet connection. HW is also pretty fast.
Deployed with docker + docker-composer using docker-compose.server.yml

If you need any additional info. please let me know.

**Comment by chrisleekr** [comment]
Hmmmmm. I have tested this feature and worked as expected. 
Let me try again later to find out whether I missed something.

**Comment by chopeta** [comment]
Thanks for the update. 
I think it was working originally when implemented and then got affected by some later-on code :-(
I have like 10 orders placed during yesterday BTC correction when the flag as disabled. In matter of fact still is disabled but I am pretty sure the bot will keep buying if Buy params. are meet (I just don't have more USDT right now so no biggy :-D)

**Comment by chrisleekr** [comment]
Wait, are you saying the bot placed multiple orders even if you already have $10 worth of BTC?

**Comment by chopeta** [comment]
> Wait, are you saying the bot placed multiple orders even if you already have $10 worth of BTC?

No, I have more than 40 symbols set in the bot. 

Yesterday I had a big part of my portafolio is USDT because I was afraid of a possible BTC correction.

To avoid the bot "buying too high" I've disable the "Buy: Trading Enabled" but for my surprise later on I cough the bot doing shopping on more than 10 different symbols :-) I was able to stop the bot, cancel a few active orders and stop the bleeding by setting "Buy: Trigger percentage" to 0.99 and let the bot running again, which early today made the bot stop working on almost all signals missing sell at least in 1 symbol (JUVUSDT), but this last part is another issue that I will report later on ;-)

Please, no complains here, just trying to explain to you what happened.

I am pretty happy with the bot results and amazed by how it works.

**Comment by chrisleekr** [comment]
40 coins? Whoa.. that is definitely not tested and not intended to use.
I am sure you understood the side effect of using 40 coins, right? Each coin will be monitored every 40 secs.

Btw, nice use of trigger percentage :)

So basically what you did is to update global settings by disabling "Buy: Trading Enabled", not symbol configuration per the coin.

But the bot purchased the coins, right?

**Comment by chopeta** [comment]
I've just learned that 40 secs thing in my scenario by reading #52 🤒 hahahaha

We do have to find a way of improving this... will try to came up with something or evaluate how to assist with the proposal described in #52 

I could swear that trigger % was the default when you first install the bot... but later on I figured that I was likely a mistake caused by a mouse scrolling using the mouse scroll button on the mouse when on top that field 🤕 very easy to reproduce 🤕 - and that lead me to the other issue I have to report, which is likely the root cause of #32, that is when you use percentages that make order creation fails on Binance side due to invalid/unacceptable amounts, I think the bot somehow crashes for that symbol and stop processing it all together until your reboot the app. I think this can be easily evitable by adding some validation and warnings on the settings GUI so user won't make mistakes when defining the % like I did on my day one and day two... inadvertently first and later on day two by not understanding how the bot works in the first plate (yes, I've set wrong % in the sell section which was cause the same problem of bot stop working - it's actually very stupid but I am pretty sure it happens to other users as well).

And yes, I've disable the "Buy: Trading Enabled" in global and had only 2 symbos in "customised" mode.

The bot is a very avid shopper hahahahaha lot of purchases yesterday...

**Comment by chopeta** [comment]
Can you point me out to which source code file is the buy order creation? Want to try and catch this bug...

**Comment by ienthach** [comment]
I confirm. The bot still open buy order with global uncheck.

**Comment by chopeta** [comment]
> I confirm. The bot still open buy order with global uncheck.

Thanks for that. At least now I know I am not crazy or something...

**Comment by chrisleekr** [comment]
> Can you point me out to which source code file is the buy order creation? Want to try and catch this bug...

So basically, there are steps for each tick.

- List of steps: https://github.com/chrisleekr/binance-trading-bot/blob/master/app/jobs/trailingTrade.js#L42

One of the steps is to determine action which decide to buy or sell: https://github.com/chrisleekr/binance-trading-bot/blob/master/app/jobs/trailingTrade/step/determine-action.js

**Comment by chrisleekr** [comment]
> I confirm. The bot still open buy order with global uncheck.

So after you update the global configuration to disable buy, but you did not update the symbol configuration to disable buy, this is correct right?

**Comment by ienthach** [comment]
> > I confirm. The bot still open buy order with global uncheck.
> 
> So after you update the global configuration to disable buy, but you did not update the symbol configuration to disable buy. Right?

My sumbols setting allway follow global setting. In UI the buy toggle in symbols is off and the label “trading is disabled” showed.

**Comment by chopeta** [comment]
> > Can you point me out to which source code file is the buy order creation? Want to try and catch this bug...
> 
> So basically, there are steps for each tick.
> 
> * List of steps: https://github.com/chrisleekr/binance-trading-bot/blob/master/app/jobs/trailingTrade.js#L42
> 
> One of the steps is to determine action which decide to buy or sell: https://github.com/chrisleekr/binance-trading-bot/blob/master/app/jobs/trailingTrade/step/determine-action.js

Thanks for clarifying... 

I was checking this:

https://github.com/chrisleekr/binance-trading-bot/blob/1a308bbb00441e5ee0297e4cd80b437bc5b8920f/app/jobs/trailingTrade/step/place-buy-order.js#L26

There's at least a misspelling in there in the word "enabled" but more than that, I don't see where's the global config is getting read here...

More over, if I am understanding the code correctly, the check of the global flag for trading would be done in 
https://github.com/chrisleekr/binance-trading-bot/blob/1a308bbb00441e5ee0297e4cd80b437bc5b8920f/app/jobs/trailingTrade/step/determine-action.js
And I see no reference for trading enabled in there...

So it looks like the bot is respecting the trading enabled flag for symbols but not globally.

Does that make sense?

**Comment by ienthach** [comment]
https://user-images.githubusercontent.com/1300386/112558499-7e4cb580-8e01-11eb-871d-9ecd8608adb0.mp4

I record video for the bug. :)

**Comment by chopeta** [comment]
Fantastic the video, what did you use to record? AwesomeScreenshot?

Imagine this happening all over the place for more than 10 symbols at time white BTC was plumbing hahahahahahaha

**Comment by ienthach** [comment]
i user macOS, i just combine "Shift + command + 5" it will record for the screen. :D

**Comment by chopeta** [comment]
> i user macOS, i just combine "Shift + command + 5" it will record for the screen. :D

More than one friend told me to get rid of my windows machines hahahahaha

**Comment by chopeta** [comment]
> i user macOS, i just combine "Shift + command + 5" it will record for the screen. :D
> 
> you are lucky guy.

Not lucky, I didn't want the buys to happen. The bot made the buys when the prices were still "too high"... I was planning on enable trading and catching the dip after BTC was lower than 52K. Of course, easy to say that now hahaha

Anyway, is part of using software... bugs will always be there.

Appreciate Chris help in getting then fixed which is all that matters.

**Comment by chopeta** [comment]
Still have a lot to learn and discover in regards to the configuration params. 
Find the best combinations for each symbol or for most of my symbols.
Will get there eventually, in the meanwhile would be nice for the bot to respect the global flags hahahaha

**Comment by chrisleekr** [comment]
> I record video for the bug. :)

Thanks for the video. Definitely, something is going on. 

As you can see from the test - https://github.com/chrisleekr/binance-trading-bot/blob/master/app/jobs/trailingTrade/step/__tests__/place-buy-order.test.js#L220

The buy order should not be placed if the trading is disabled, which the configuration is the symbol configuration.

So possible bug in the global/symbol configuration setting.
Anyway, let me fix up the configuration issue first.

**Comment by chopeta** [comment]
Just as disclaimer for other people that may read this later on and possibly get wrong ideas...

If you don't want to risky your money, you should really test this bot first for weeks using test mode, and only after that go live. 

And do it again every time a new release is published and you plan to upgrade, because at the end of day with you playing with money here.

Be careful and DYOR!!!!

**Comment by chrisleekr** [comment]
Oh -_- I just found typo...... `enabed` ...

**Comment by chopeta** [comment]
> Oh -_- I just found typo...... `enabed` ... Damn...

Told you this in one of my prev. comments...

Can I be hired already for the Q&A department?

hahahahaha

Jokes a part, I am glad we're moving forward.

**Comment by chrisleekr** [comment]
> Told you this in one of my prev. comments...

I saw your comment, but I didn't know where. :)

Thanks guys.

**Comment by chrisleekr** [comment]
@chopeta @ienthach 

I have released a new version that contains the fix for the global configuration.
As well as a typo that caused the issue with buying disabled.

https://github.com/chrisleekr/binance-trading-bot/releases/tag/v0.0.65

Please close the issue if you think it is fixed.

**Comment by ienthach** [comment]
im checking on this. thank you.

**Comment by chopeta** [comment]
Will deploy now... How do I upgrade in docker?

**Comment by ienthach** [comment]
I do stop docker-compose stop. git pull then start again

**Comment by chrisleekr** [comment]
If you are using `docker-compose.server.yml`, then simply

```
docker-compose -f docker-compose.server.yml pull
docker-compose -f docker-compose.server.yml up -d
```

**Comment by chopeta** [comment]
Thanks for the help with docker guys. Rookie here.
I had to stop, remove the image and then start again.
Just stop and start was not getting the image updated... git pull was the first thing I did.
Anyway, looks like I am on the latest version now, reported #59 just to be sure :-)
Let's see how it goes now...
Thanks for the fixes @chrisleekr, appreciate it!

**Comment by ienthach** [comment]
i confirms the bug if fixed.

<img width="476" alt="Screen Shot 2021-03-27 at 14 46 36" src="https://user-images.githubusercontent.com/1300386/112714438-f94ec280-8f0c-11eb-81ac-4f17cfa078be.png">

**Comment by chopeta** [comment]
Working here as expected. 
Thanks for the fix @chrisleekr
